design-docs/opmodes.md

```diff
@@ -0,0 +1,526 @@
+
+The drop-down lists have the opmodes (as defined by user code) organized into groups, similar to the below.  The groups and opmodes are sorted.
+
+![Combo box with groups](https://i.sstatic.net/4Cmfk.gif)
+
+Only the opmodes appropriate to the selected mode are shown in the corresponding list.  For example, if the "Test" option is selected, a single drop-down list with only test opmodes listed will be shown.
+
+## Robot Code User-Facing API
+
+Java is used for illustrative purposes.  Python and C++ should follow a similar structure, except that C++ doesn't support annotations.
+
+### RobotBase
+
+The `RobotBase` class is the base class for the user's `Robot` class.  It also implements the private library machinery for robot startup and robot execution (including creating and transitioning between opmodes in accordance with the opmode lifecycle, as described in the following section).
+
+```java
+public abstract class RobotBase {
+  public void disabledStart() {
+    // this code is called when the robot enters disabled state (including at startup)
+  }
+
+  public void disabledPeriodic() {
+    // this code is called periodically while the robot is disabled
+  }
+
+  public void disabledEnd() {
+    // this code is called when the robot exits disabled state, prior to a opmode starting
+  }
+
+  // these functions allow users to add opmodes without annotations
+  public void addAutonomousOpMode(Supplier<OpMode> factory, String name, String group, String description) {...}
+  public void addTeleoperatedOpMode(Supplier<OpMode> factory, String name, String group, String description) {...}
+  public void addTestOpMode(Supplier<OpMode> factory, String name, String group, String description) {...}
+}
+```
+
+### OpMode Classes
+
+There are library base classes for different coding styles of routines.  The `OpMode` interface serves as the base interface for all opmodes.  Most users will use either `LinearOpMode` or `PeriodicOpMode` abstract base class implementations of this interface, but this interface enables users to create more customized opmodes while still utilizing the core robot base class opmode-switching implementation.
+
+The full lifecycle of a opmode is as follows:
+- Operator selects opmode on DS -> opmode object is constructed
+- If different opmode is selected -> `close()` is called, object is released to GC
+- While opmode is selected and robot is disabled, `disabledPeriodic()` is called
+- When the robot is enabled in the selected opmode, `opmodeRun()` is called; the result of this is different for different opmode base classes:
+  - `start()` is called once (for both `LinearOpMode` and `PeriodicOpMode`)
+  - `run()` is called (for `LinearOpMode`), or `periodic()` is called periodically (for `PeriodicOpMode`)
+- When the robot is disabled, `opmodeStop()` is called (which results in `end()` being called for both `LinearOpMode` and `PeriodicOpMode`), followed by `close()`, object is released to GC
```

Comment:
I'm not entirely sure how we want to reword this, but this needs a little bit of updating:
* `OpMode.close()` was replaced with `OpMode.opmodeClose()`
* `LinearOpMode.opmodeRun()` calls `end()` after `run()` finishes
* `LinearOpMode.opmodeStop()` results in `LinearOpMode.isRunning()` returning false, which is then up to the user to make `run()` finish so that `end()` can be run